### PR TITLE
Resolve #11727

### LINKED ISSUE #11727: BuildServiceProvider in ConfigureServices should raise a warning
### Is your feature request related to a problem? Please describe.
To encourage better practices, it would be helpful for an Analyzer to raise a warning if BuildServiceProvider is called in ConfigureServices.  This is an antipattern we should discourage

### PR DESCRIPTION
**Pull request by wu-yafeng**
Summary of the changes (Less than 80 chars)
 - Add a BuildServiceProviderValidator for checking startup method 'ConfigureServices(IServiceCollection,..)'
 - Detail 2

Addresses #11727

### CODE REVIEW

**Comment by wu-yafeng** [comment]
It seems no RPs for this issue.So,I do it~ :smile:

File: src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs [hunk 1]
```diff
@@ -16,5 +16,15 @@ public partial class StartupAnalzyer : DiagnosticAnalyzer
...
             helpLinkUri: "https://aka.ms/YJggeFn");
+
+        internal readonly static DiagnosticDescriptor BuildServiceProviderShouldNotCalledInConfigureServicesMethod = new DiagnosticDescriptor(
+            "MVC1007",
+            "Cannot use BuildServiceProvider in ConfigureServices.",
+            "BuildServiceProvider called by runtime,don't call this method in your application code directly",
+            "Design",
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            helpLinkUri: "https://github.com/aspnet/AspNetCore/issues/11727"
```
Comment (pranavkm) [inline src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs]: @davidfowl suggestions on where this should point to?

Comment (davidfowl) [reply src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs]: I would assume these should all point to a single docs article via an aka.ms link

Comment (rynowak) [reply src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs]: This should be an aka.ms and it should point to a docs article.

Comment (rynowak) [reply src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs]: @wu-yafeng - we'll get back to you with a URL to put here

**Comment by Tratcher** [comment]
Why are Startup analyzers in MVC?

**Comment by rynowak** [comment]
There's a work item on my plate to get the more general `Microsoft.AspNetCore.Analzyers` assembly shipped in the SDK. It's a purely point in time issue.

File: src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs [hunk 2]
```diff
@@ -16,5 +16,15 @@ public partial class StartupAnalzyer : DiagnosticAnalyzer
             DiagnosticSeverity.Warning, 
             isEnabledByDefault: true,
             helpLinkUri: "https://aka.ms/YJggeFn");
+
+        internal readonly static DiagnosticDescriptor BuildServiceProviderShouldNotCalledInConfigureServicesMethod = new DiagnosticDescriptor(
+            "MVC1007",
+            "Cannot use BuildServiceProvider in ConfigureServices.",
```
Comment (rynowak) [inline src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs]: Suggestion: `Do not call 'IServiceCollection.BuildServiceProvider' in 'ConfigureServices'.`

File: src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs [hunk 3]
```diff
@@ -16,5 +16,15 @@ public partial class StartupAnalzyer : DiagnosticAnalyzer
             DiagnosticSeverity.Warning, 
             isEnabledByDefault: true,
             helpLinkUri: "https://aka.ms/YJggeFn");
+
+        internal readonly static DiagnosticDescriptor BuildServiceProviderShouldNotCalledInConfigureServicesMethod = new DiagnosticDescriptor(
+            "MVC1007",
```
Comment (rynowak) [inline src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs]: @Tratcher - raised an interesting point. Right now these build from inside the MVC solution - I'm planning to move these to a more general place, and I'll handle the update of this ID when I do it.

File: src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs [hunk 4]
```diff
@@ -16,5 +16,15 @@ public partial class StartupAnalzyer : DiagnosticAnalyzer
...
+            "Cannot use BuildServiceProvider in ConfigureServices.",
+            "BuildServiceProvider called by runtime,don't call this method in your application code directly",
```
Comment (rynowak) [inline src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs]: Suggestion: `Calling 'BuildServiceProvider' from application code results in an additional copy of singleton services being created. Consider alternatives such as dependency injecting services as parameters to `Configure`.

The docs article should cover:
- DI of parameters to configure
- Using DI with options https://andrewlock.net/access-services-inside-options-and-startup-using-configureoptions/
- @davidfowl  anything else?

File: src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs (hunk 1 above)
Comment (rynowak) [reply src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs]: @wu-yafeng - please make this point to `https://aka.ms/AA5k895`. I'll write a docs article as part of this release.

**Review by rynowak — CHANGES_REQUESTED** [review]

**Comment by wu-yafeng** [comment]
There are no tests or asserts for `DiagnosticDescriptor` properties such as `Id`,`Title.
Should I add tests or asserts for them?

**Comment by rynowak** [comment]
> Should I add tests or asserts for them?

I don't think it's necessary, those tests would just verify that the text is what you typed in. Those kinds of tests are valuable if the diagnostic output contains format parameters.

**Review by rynowak — APPROVED** [review]

File: src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs (hunk 4 above)
Comment (davidfowl) [reply src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs]: Does this warn if you return the IServiceProvider?

Comment (wu-yafeng) [reply src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs]: No,this warn if called `BuildServiceProvider` in 'ConfigureServices'.

It seems add validators such as `ConfigureServiceMethodSignatureValidator` and `ConfigureMethodSignatureValidator` to check startup methods signatures

Comment (davidfowl) [reply src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs]: This only affects 3.0 applications right?

Comment (rynowak) [reply src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs]: These analyzers will only show up for 3.0+ because they ship in the SDK.

**Comment by rynowak** [comment]
Thanks!

**Pull request merged**